src/modules/auth/components/RoleDropdown.js

```diff
@@ -0,0 +1,164 @@
+	UncontrolledTooltip } from "reactstrap";
+
+const RoleDropdown = React.memo(({props, tenantObj, selectedTenants, setSelectedTenants, idx}) => {
+
+	const { t } = useTranslation();
+	const [dropdownOpen, setDropdownOpen] = useState(false);
+	const [displayTenantRoles, setDisplayTenantRoles] = useState({data: []});
+	const [count, setCount] = useState(0);
+	const [filter, setFilter] = useState("");
+	const [limit, setLimit] = useState(5);
+	const [items, setItems] = useState(0);
+
+	let SeaCatAuthAPI = props.app.axiosCreate('seacat_auth');
+	let message;
+
+	useEffect(() => {
+		retrieveRoleList(tenantObj);
+	}, [limit]);
+
+	// TODO: this useEffect should match selected roles with the ones we want to display and re-add the ones which were removed from selected roles in BulkAssignmentContainer
+	useEffect(() => {
+		if (tenantObj.selectedRole && (tenantObj?.selectedRole?.length > 0)) {
+			let display = {...displayTenantRoles};
+			// newDisplayData will hold roles coming from the service minus the ones we've already assigned
+			let newDisplayData = [];
+			// selectedTenants[idx]?.roles?.data.map((tenantRole) => {
+			tenantObj?.roles?.data.map((tenantRole) => {
+				// we are comparing already selected roles with all roles and keeping only the elements, which are not included in selectedRole(s)
+				let found = tenantObj?.selectedRole.find(el => el === tenantRole._id);
+				if (!found) {
+					newDisplayData.push(tenantRole);
+				}
+			});
+			display['data'] = newDisplayData;
+			setDisplayTenantRoles(display);
+		}
+		if (tenantObj.selectedRole && tenantObj?.selectedRole?.length === 0) {
+			setDisplayTenantRoles(tenantObj.roles);
+		}
+	}, [selectedTenants, items]);
+
+// uncomment, when search functionality in roles is enabled on the backend
+	// // sets 0.5s delay before triggering the search call when filtering through tennants
+	// useEffect(() => {
+	// 	if (timeoutRef.current !== null) {
+	// 		clearTimeout(timeoutRef.current);
+	// 	}
+	// 	timeoutRef.current = setTimeout(() => {
+	// 		timeoutRef.current = null;
+	// 		retrieveRoleList(tenantObj);
+	// 	}, 500);
+	// }, [filter]);
+
+	// const handleFilter = (e) => {
+	// 	setLimit(5);
+	// 	setFilter(e.target.value);
+	// }
+
+	// fetch roles for Tenant dropdowns
+	const retrieveRoleList = async (tenantObj) => {
+		let response;
+		let id = tenantObj._id;
+		let parameters = {f: filter, i: limit, exclude_global: true};
+		if (tenantObj._id === "Global roles") {
+			id = '*';
+			parameters['exclude_global'] = false;
+		}
+		try {
+			response = await SeaCatAuthAPI.get(`/role/${id}`, {params: parameters});
+			tenantObj['roles'] = response.data;
+			setCount(response.data.count);
+			/* this Items state was created just as a helper to cause triggering of useEffect
+			on line 23 when new roles were triggered after pressing the more button in role dropdown.
+			This is not the most elegant solution, I'm open for updgrade */
+			setItems(response?.data?.data.length);
```

Comment:
@petrKavulok so then you can use tenantObj as a trigger for useEffect on line 25 (btw. dont write line numbers in the code, since it will change and your comment will be out of date very soon as it happened already)

why I suggested to use tenantObj as a trigger? its because you are modifying it within try catch anyway, so you dont need to use any special state here I suppose